📦 NEW: Codelab Cloudinary URL API (#1041)
GoogleChrome/web.dev: src/site/content/en/fast/codelab-cloudinary-url-api/index.md

    <source src="./cloudname.mp4" type="video/mp4; codecs=h264">
  </video>
  <figcaption class="w-figcaption">
    Setting a custom Cloud Name.
  </figcaption>
</figure>

## The Cloudinary Fetch URL

The [Cloudinary Fetch URL](https://cloudinary.com/documentation/fetch_remote_images) feature is a quick way to deliver images from remote URLs. The images can be manipulated and optimized on-the-fly, before being cached and delivered through fast, localized CDNs and not via local web servers. It requires no build step and no manual upload.

In this code lab you will replace the image links with Fetch URLs that will automagically upload, transform, and optimize your images.

For example:

```html/1/0
<img src="https://codelab-cloudinary.glitch.me/images/flower1.png"/>
<img src="https://res.cloudinary.com/demo/image/fetch/q_auto,f_auto/https://codelab-cloudinary.glitch.me/images/flower1.png"/>
```

## Optimize an image

- Click the **Remix to Edit** button (in the top-right corner) to make the project editable.
- In the next few steps, you will replace the image link on Line #25 of `index.html` file, with a fetch URL.

### Create a Fetch URL

> To create a fetch URL, you have to prepend the following prefix to the existing URL of the image.

Reviewer: ```suggestion
To create a fetch URL, prepend the following prefix to the existing URL of the image:
```

> An example fetch URL looks like this:

Reviewer: ```suggestion
```

Now replace the `<remote_image_url>` with the original URL for the first image. The final URL looks like this:

```html/1/0
https://res.cloudinary.com/demo/image/fetch/q_auto,f_auto/<remote_image_url>
https://res.cloudinary.com/demo/image/fetch/q_auto,f_auto/https://codelab-cloudinary.glitch.me/images/flower1.png
```

### Final image URL

The final image URL after replacing `<cloud_name>`, `<transformations>`, and `<remote_image_url>` looks like this:

```html
https://res.cloudinary.com/demo/image/fetch/q_auto,f_auto/https://codelab-cloudinary.glitch.me/images/flower1.png

```

This works for other images too. You can auto-optimize any image by prepending the following to its URL:

```html
https://res.cloudinary.com/demo/image/fetch/q_auto,f_auto/
```

### Results

After updating the `src` attribute of the first image (line #25) with the fetch URL, your HTML should look like this:

```html/2/1
<div class="wrapper">
  <img src="https://codelab-cloudinary.glitch.me/images/flower1.png"/>
  <img src="https://res.cloudinary.com/demo/image/fetch/q_auto,f_auto/https://codelab-cloudinary.glitch.me/images/flower1.png"/>
  <div class="price">Violet bouquet- $9</div>
</div>
```

> This helps save more than 90% on image size.

Reviewer: ```suggestion
This decreases image size by more than 90%.
```

        </tr>
      </thead>
    </tbody>
  </table>
</div>

The photo on the right is 92.39% smaller than the one on the left, yet would probably look identical to the average user.

{% Aside 'objective' %}
  Now prepend all image links in the `index.html` with `https://res.cloudinary.com/demo/image/fetch/q_auto,f_auto/`. Make sure to change `demo` to your `cloud_name`.
{% endAside %}

## ✔︎ Check-in

Your `index.html` file should now look like this:

```html/(1,1,5,9,13,17,21)
<div class="wrapper">
  <img src="https://res.cloudinary.com/demo/image/fetch/q_auto,f_auto/https://codelab-cloudinary.glitch.me/images/flower1.png" alt="Yellow bouquet" />
  <div class="price">Yellow bouquet - $9</div>
</div>
<div class="wrapper">
  <img src="https://res.cloudinary.com/demo/image/fetch/q_auto,f_auto/https://codelab-cloudinary.glitch.me/images/flower2.jpg" alt="Cream bouquet" />
  <div class="price">Cream bouquet - $5</div>
</div>
<div class="wrapper">
  <img src="https://res.cloudinary.com/demo/image/fetch/q_auto,f_auto/https://codelab-cloudinary.glitch.me/images/flower3.png" alt="Light pink" />
  <div class="price">Light pink bouquet - $6</div>
</div>
```

## Image Performance Gains

> There are impressive web performance gains by using <a href="https://cloudinary.com/documentation/image_optimization">cloud to optimize images</a> without losing quality.

Reviewer: ```suggestion
These are the results of using <a href="https://cloudinary.com/documentation/image_optimization">cloud to optimize images</a> without losing quality.
```

        <th>SIZE</th>
      </tr>
      <tr>
        <td>flower1.png</td>
        <td>289 KB</td>
        <td>22 KB</td>
        <td>↓ -92.39%</td>
      </tr>
      <tr>
        <td>flower2.jpg</td>
        <td>59 KB</td>
        <td>19 KB</td>
        <td>↓ -67.8%</td>
      </tr>
      <tr>
        <td>flower3.png</td>
        <td>367 KB</td>
        <td>38 KB</td>
        <td>↓ -89.65%</td>
      </tr>
    </tbody>
  </table>
</div>

Hooray! These results are much better.

## Lighthouse Audit

Lastly, it's a good idea to use <a target="_blank" rel="noopener noreferrer" href="https://web.dev/discover-performance-opportunities-with-lighthouse">Lighthouse</a> to verify the changes that you just made.

Lighthouse's "Efficiently encode images" performance audit can let you know if
the JPEG images on your page are optimally compressed.

- To preview the site press the **Share** button (in the right-bottom corner) and copy the **Live App** link.
- Run the Lighthouse performance audit (Lighthouse ❯ Options ❯ Performance) on
  the live version of your Glitch and verify that the "Efficiently encode
  images" as well as "Serve images in next-gen formats" audits were passed.

### Results

> Lighthouse audit results for when images are optimized.

Reviewer: ```suggestion
```

Reviewer: Captions should be sufficient.

<figure class="w-figure">
  <img class="w-screenshot" src="./optimized-images.png" alt="Images Not Optimized">
  <figcaption class="w-figcaption w-figcaption--fullbleed">
    Fig. 2 — Results with optimized images.
  </figcaption>
</figure>

> If you've optimized your images correctly the Lighthouse audit should have a perfect score. As you can see the page load time is down by 4.05 secs. That is impressive.

Reviewer: ```suggestion
If you've optimized your images correctly the Lighthouse audit should have a perfect score. As you can see First Meaningful Paint is down by 2.5 secs. That is impressive.
```

Reviewer: (Changing this to match the image, which highlights First Meaningful Paint.)

> _Lighthouse audit results when images are not optimized._

Reviewer: ```suggestion
```

authors:
  - ahmadawais
date: 2019-06-24
description: |
  In this codelab, learn how to use Cloudinary with its URL API to optimize JPEG and
  PNG images for faster download and serve next-gen formats.
glitch: codelab-cloudinary
path: index.html:25:0
related_post: use-imagemin-to-compress-images
---

In this codelab, learn how to use a simple URL API to optimize images for faster downloads and serve media in next-gen formats. All of that happens dynamically in the cloud without any build step.

## Get a Cloudinary account

Click [here](https://cloudinary.com/users/register/free) to sign up for a free Cloudinary account. Remember to set a custom cloud name (at the end of the registration form). Your cloud name is used to build the URL that your images will be publicly available from.

<figure class="w-figure">
  <video controls autoplay loop muted class="w-screenshot">
    <source src="./cloudname.webm" type="video/webm; codecs=vp8">
    <source src="./cloudname.mp4" type="video/mp4; codecs=h264">
  </video>
  <figcaption class="w-figcaption">
    Setting a custom Cloud Name.
  </figcaption>
</figure>

## The Cloudinary Fetch URL

> This codelab demonstrates how you can use the power of a cloud to optimize your images and serve next-gen formats dynamically. You will learn how to use dynamic cloud Fetch and Image Transformations features. You will replace the image links with Fetch URLs (explained below) that will automagically upload, transform, and optimize your images.

Author: Done in 761ec21.

2. [**Format**](https://cloudinary.com/documentation/image_transformation_reference#format_parameter): The parameter `f` can automatically convert delivered images to WebP and JPEG-XR on supported browsers to save bandwidth and optimize delivery time. Again use `f_auto` to auto format the image.

### #3. Replace `<remote_image_url>`

Now replace the `<remote_image_url>` with the original link of the first image. The final URL looks like this:

```html/1/0
https://res.cloudinary.com/demo/image/fetch/q_auto,f_auto/<remote_image_url>
https://res.cloudinary.com/demo/image/fetch/q_auto,f_auto/https://codelab-cloudinary.glitch.me/images/flower1.png
```

### Final image URL

The final image URL after replacing `<cloud_name>`, `<transformations>`, and `<remote_image_url>` looks like this:

```html
https://res.cloudinary.com/demo/image/fetch/q_auto,f_auto/https://codelab-cloudinary.glitch.me/images/flower1.png

```

> Which means that you can auto-optimize any image by prepending the following to its URL.

Reviewer: ```suggestion
This works for other images too. You can auto-optimize any image by prepending the following to its URL:
```

---
layout: codelab
title: Optimize images and serve next-gen formats using Cloudinary
authors:
  - ahmadawais
date: 2019-06-24
description: |
  In this codelab, learn how to use Cloudinary with its URL API to optimize JPEG and
  PNG images for faster download and serve next-gen formats.

> glitch: codelab-cloudinary?path=index.html:25:0

Author: I got the codelab reviewed at a local meetup. 8 out of 11 attendees had problems figuring out how to open `index.html` since it was there first time with Glitch and small screens glitch hides the left Files Explorer. 

Then I worked around a way for the codelab to open up with the exact file instead `index.html`. This helped them get started quickly.

So, I'd like to request to keep `?path=index.html:25:0` part there for better Developer Experience.

Thoughts?

> glitch: codelab-cloudinary?path=index.html:25:0

Reviewer: That sounds reasonable to me.

@robdodson - Can you confirm that adding URL params to the glitch name won't break anything?

> glitch: codelab-cloudinary?path=index.html:25:0

Reviewer 2: I think you can use the path param in YAML to do this:

```
glitch: codelab-cloudinary
path: index.html:25:0
```

> glitch: codelab-cloudinary?path=index.html:25:0

Author: @robdodson thanks for sharing that. Fixed in db09fa3

related_post: use-imagemin-to-compress-images
---

In this codelab, learn how to use a simple URL API to optimize images for faster downloads and serve media in next-gen formats. All of that happens dynamically in the cloud without any build step.

## Get a Cloudinary account

Click [here](https://cloudinary.com/users/register/free) to sign up for a free [Cloudinary account](https://cloudinary.com/users/register/free). Remember to set a custom cloud name (at the end of the registration form). Your cloud name is used to build the URL that your images will be publicly available from.

For example: `https://res.cloudinary.com/`<span class="w-c--success">my_cloud_name</span>`/...`

<figure class="w-figure">
  <a href="https://cloudinary.com/users/register/free" target="_blank" rel="noopener noreferrer">

>     <img class="w-screenshot" src="./cloudname.gif" alt="Setting a custom Cloud Name" />

Author: On the meetup, I found out that many users would skip looking/setting up a custom `cloudName`. It's something we are working on at Cloudinary. But once I added this GIF it made them set it up easily.

I have optimized this GIF to be smaller in size than other images. It's 76 KB only. So, my question is can we keep it? 

Let me know and I'm game either way!

>     <img class="w-screenshot" src="./cloudname.gif" alt="Setting a custom Cloud Name" />

Reviewer 3: As a rule, we favor video over GIFs: https://web.dev/handbook/markup-media/#video-hosted-on-web.dev

>     <img class="w-screenshot" src="./cloudname.gif" alt="Setting a custom Cloud Name" />

Author: @mfriesenhahn Sure thing. I have uploaded the [video here](https://github.com/GoogleChrome/web.dev/blob/0ba6faf3e7d73595d4a95727a7de326701e673b3/src/site/content/en/fast/codelab-cloudinary-url-api/coudname.mp4). 

>Once you have a video ready, message your content reviewer, and they'll help you upload it to the web.dev Google Cloud Storage bucket.

I hope you can take care of this now. :)

2. [**Format**](https://cloudinary.com/documentation/image_transformation_reference#format_parameter): The parameter `f` can automatically convert delivered images to WebP and JPEG-XR on supported browsers to save bandwidth and optimize delivery time. Again use `f_auto` to auto format the image.

### #3 REPLACE `<remote_image_url>`

Now replace the `<remote_image_url>` with the original link of the first image. The final URL looks like this:

```html/1/0
https://res.cloudinary.com/demo/image/fetch/q_auto,f_auto/<remote_image_url>
https://res.cloudinary.com/demo/image/fetch/q_auto,f_auto/https://codelab-cloudinary.glitch.me/images/flower1.png
```

### Final image URL

The final image URL after replacing `<cloud_name>`, `<transformations>`, and `<remote_image_url>` looks like this:

```html
https://res.cloudinary.com/demo/image/fetch/q_auto,f_auto/https://codelab-cloudinary.glitch.me/images/flower1.png

```

Which means that you can auto-optimize any image by prepending the following to its URL.

```html
https://res.cloudinary.com/demo/image/fetch/q_auto,f_auto/
```

> ### RESULTS

Reviewer: ```suggestion
### Results
```

> After replacing the link of the first image (line #25) with the fetch URL, your HTML should look like this:

Reviewer: ```suggestion
After updating the `src` attribute of the first image (line #25) with the fetch URL, your HTML should look like this:
```

                ><img src="https://codelab-cloudinary.glitch.me/images/flower1.png" width="100%"
              /></a>
            </p>
            <strong>289 KB</strong> (Original)
          </th>
          <th>
            <p>
              <a
                href="https://res.cloudinary.com/demo/image/fetch/q_auto,f_auto/https://codelab-cloudinary.glitch.me/images/flower1.png"
                target="_blank"
                rel="noopener noreferrer"
              >
                <img
                  src="https://res.cloudinary.com/demo/image/fetch/q_auto,f_auto,a_ignore/https://codelab-cloudinary.glitch.me/images/flower1.png"
                  width="100%"
                />
              </a>
            </p>
            <strong>22 KB</strong> (Cloudinary)
          </th>
        </tr>
      </thead>
    </tbody>
  </table>
</div>

The photo on the right is 92.39% smaller than the one on the left, yet would probably look identical to the average user.

{% Aside 'objective' %}
  Now prepend all image links in the `index.html` with `https://res.cloudinary.com/demo/image/fetch/q_auto,f_auto/`. Make sure to change `demo` to your `cloud_name`.
{% endAside %}

## ✔︎ Check-in

> Your `index.html` file should now look like this:

Author: Sure thing!

> Your `index.html` file should now look like this:

Author: 🐛 FIX: in 399812f

```html
<!--
Copyright 2018 Google LLC.
SPDX-License-Identifier: Apache-2.0
-->
<!DOCTYPE html>
<html lang="en">
  <head>
    <title>Cloudinary URL API Demo: Optimized Images</title>
    <meta name="Description" content="In this codelab, learn how to use Cloudinary with its simple URL API to optimize images for faster downloads and serve media in next-gen formats. All of that happens dynamically in the Cloud without any build step.">
    <link href="https://cloudinary-res.cloudinary.com/image/asset/favicon-192x192-d6a96e11dd5adfebbb717d154665ee80.png" rel="icon" sizes="192x192" type="image/png">
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style>body {text-align: center}img {border: 1px solid black}.wrapper {margin-bottom: 24px}.wrapper .price {font-size: 24px}</style>
  </head>
  <body>
    <header>
      <h1>
        Foobar Flowershop
      </h1>
      <p>We sell the best flowers!</p>
    </header>
    <div class="wrapper">

>       <img src="https://res.cloudinary.com/demo/image/fetch/q_auto,f_auto/https://cloudinary-url-api.glitch.me/images/flower1.png" alt="Yellow bouquet" />

Reviewer: cloudinary-url-api -> codelab-cloudinary

Author: Ah dang it. Sorry about this. Last minute glitch name change, I thought I fixed all of these.

Author: 🐛 FIX: 0988735

>       <div class="price">Yellow bouquet - $9</div>

Reviewer: I think I'd make the glitch slightly shorter (maybe only 3 or 4 flowers?). This cuts down on the amount of changes people need to make the HTML & makes this code snippet shorter too.

Author: 👌 IMPROVE: Shorten to Three images → d4040af

    <div class="wrapper">
      <img src="https://res.cloudinary.com/demo/image/fetch/q_auto,f_auto/https://cloudinary-url-api.glitch.me/images/flower3.png" alt="Light pink" />
      <div class="price">Light pink bouquet - $6</div>
    </div>
    <div class="wrapper">
      <img src="https://res.cloudinary.com/demo/image/fetch/q_auto,f_auto/https://cloudinary-url-api.glitch.me/images/flower4.jpg" alt="Violet bouquet" />
      <div class="price">Violet bouquet - $8</div>
    </div>
    <div class="wrapper">
      <img src="https://res.cloudinary.com/demo/image/fetch/q_auto,f_auto/https://cloudinary-url-api.glitch.me/images/flower5.jpg" alt="Polka dot" />
      <div class="price">Polka dot bouquet - $11</div>
    </div>
    <div class="wrapper">
      <img src="https://res.cloudinary.com/demo/image/fetch/q_auto,f_auto/https://cloudinary-url-api.glitch.me/images/flower6.jpg" alt="Burgandy bouquet" />
      <div class="price">Burgandy bouquet - $7</div>
    </div>
  </body>
</html>
```

## Image Performance Gains

There are impressive web performance gains by using <a href="https://cloudinary.com/documentation/image_optimization">cloud to optimize images</a> without losing quality.

> <div class="w-table-wrapper">

Author: Glad you liked it. Fun comparison. 

P.S. I use [Numi](https://numi.io/) mac calculator to calculate the optimized image sizes.

P.P.S In case you're into this, I've written a custom extension for Numi calculator where I only have to write `optim(newSize;oldSize)` of an image to calculate the percentage savings.

![optim](https://on.ahmda.ws/29b9a9/c)

Here's the code ↓

```js
/**
 * Optimization percentage in new vs old size.
 *
 * @usage optim(newSize;oldSize)
 * @usage optim(10;100)
 * @param Number newSize The new reduced size.
 * @param Number oldSize The old large size.
 * @author Ahmad Awais <https://twitter.com/MrAhmadAwais>
 */

numi.addFunction({ id: 'optim', phrases: 'optim' }, function(values) {
	return {
		double: (values[0].double / values[1].double) * 100 - 100
	};
});
```

        <td><span class="w-c--success">47 KB</span></td>
        <td><span class="w-c--success">↓ -51.04%</span></td>
      </tr>
      <tr>
        <td>flower6.jpg</td>
        <td><span class="w-c--warning">68 KB</span></td>
        <td><span class="w-c--success">24 KB</span></td>
        <td><span class="w-c--success">↓ -64.71%</span></td>
      </tr>
    </tbody>
  </table>
</div>

Hooray! These results are much better.

## Lighthouse Audit

Lastly, it's a good idea to use <a target="_blank" rel="noopener noreferrer" href="https://web.dev/discover-performance-opportunities-with-lighthouse">Lighthouse</a> to verify the changes that you just
made.

Lighthouse's "Efficiently encode images" performance audit can let you know if
the JPEG images on your page are optimally compressed.

- To preview the site press the **View App** button (in the right-bottom corner).
- Run the Lighthouse performance audit (Lighthouse ❯ Options ❯ Performance) on
  the <a target="_blank" rel="noopener noreferrer" href="https://codelab-cloudinary.glitch.me/">live version</a> of your Glitch and verify that the "Efficiently encode
  images" as well as "Serve images in next-gen formats" audits were passed.

> ### RESULTS

Reviewer: I think the "Results" section can be omitted.

Author: Should I delete this entire section?

![img](https://on.ahmda.ws/2104a7/c)

Reviewer: After re-reading this a bunch of times I've changed my mind. I think this would make more sense to me:

- Keep the "Results" section - but show the optimized Lighthouse result first and the unoptimized result second.
- Remove "optmized.html" so people aren't tempted to skip through the code lab just to get to the flashy Lighthouse score :)

If I were reading this I think I'd run the Lighthouse audit when I got to the last bullet point of "Lighthouse Audit". After that I think I'd expect to see the results of a successful/optimized Lighthouse run (so I could verify that I got the same results and did everything correctly). Then, if that was followed by an image of the Lighthouse report for the unoptimized "before" version, I could see the impact of what I just did, without necessarily having to run Lighthouse again.

Author: @khempenius — That makes perfect sense. I've made the following changes:

- Removed `optimized.html` — good call.
- Moved the optimized results to the top.
- Removed the redundant info from the non-optimized results.
